### PR TITLE
Reuse Http2MessageBody

### PR DESCRIPTION
**Pull request by JamesNK**
Before:
image

After:
image

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http/ZeroContentLengthMessageBody.cs [hunk 1]
```diff
@@ -11,8 +11,8 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http
     internal sealed class ZeroContentLengthMessageBody : MessageBody
     {
         public ZeroContentLengthMessageBody(bool keepAlive)
-            : base(null)
         {
+            Reset(null);
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http/ZeroContentLengthMessageBody.cs]: Nit: this is unnecessary.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs [hunk 2]
```diff
@@ -12,13 +12,12 @@ namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2
 {
     internal sealed class Http2MessageBody : MessageBody
     {
-        private readonly Http2Stream _context;
+        private Http2Stream _context;
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Why is this necessary?

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs [hunk 3]
```diff
@@ -46,16 +45,29 @@ protected override void OnDataRead(long bytesRead)
             AddAndCheckConsumedBytes(bytesRead);
         }
 
-        public static MessageBody For(Http2Stream context)
+        public static MessageBody For(Http2Stream context, Http2MessageBody previousMessageBody)
         {
             if (context.ReceivedEmptyRequestBody)
             {
                 return ZeroContentLengthClose;
             }
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2MessageBody.cs]: Since the above logic was moved to Http2Stream.CreateMessageBody, do we need it here? I wouldn't want to accidentally  pool ZeroContentLengthClose.

**Review by halter73 — APPROVED** [review]

File: src/Servers/Kestrel/Core/src/Internal/Http/MessageBody.cs [hunk 4]
```diff
@@ -73,6 +73,17 @@ public virtual Task StopAsync()
 
         protected virtual Task OnStopAsync() => Task.CompletedTask;
 
+        public virtual void Reset()
```
Comment (Tratcher) [inline src/Servers/Kestrel/Core/src/Internal/Http/MessageBody.cs]: It would be good to call Reset from the constructor to ensure consistent state the first time it's used.
